Rdf4j update module

blcham opened the pull request.

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jUpdateModule.java:
```diff
@@ -0,0 +1,101 @@
+        for (Resource updateQuery : updateQueries) {
+            String text = updateQuery.getProperty(SP.text).getLiteral().getString();
+            String[] queries = text.split("\\}");
+            for (int j = 0; j < queries.length - 1; j++) {
```
very complicated logic, should be split into smaller functions that we would be able to understand

something like

String parseQuery() ...

blcham replied:
@rodionnv

blcham commented on s-pipes-modules/module-rdf4j/src/test/java/cz/cvut/spipes/modules/Rdf4jUpdateModuleTest.java:
```diff
@@ -0,0 +1,41 @@
+    void executeSelf() {
+        Rdf4jUpdateModule module = new Rdf4jUpdateModule();
+        ExecutionContext newContext = module.executeSelf();
+//        System.out.println(newContext.getDefaultModel().listStatements().toList());
```
should not be here

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jUpdateModule.java:
```diff
@@ -0,0 +1,101 @@
+    private static Property getParameter(final String name) {
+        return ResourceFactory.createProperty(PROPERTY_PREFIX_URI + "/" + name);
+    }
+    private String getUpdateType(String updateString){
```
Why do we need this anyway?

blcham commented on s-pipes-modules/module-rdf4j/src/test/java/cz/cvut/spipes/modules/Rdf4jDeployModuleTest.java:
```diff
@@ -12,12 +12,12 @@
     @Disabled
     public void testDeployEmpty() throws Exception {
-        final Rdf4jModule moduleRdf4j = new Rdf4jModule();
+        final Rdf4jDeployModule moduleRdf4j = new Rdf4jDeployModule();
```
we need to test it by in-memory database

blcham commented on s-pipes-modules/module-rdf4j/src/test/java/cz/cvut/spipes/modules/Rdf4jUpdateModuleTest.java:
```diff
@@ -0,0 +1,41 @@
+
+        moduleRdf4j.setConfigurationResource(root);
+
+        // TODO: currently running server is needed;
```
we need to do testing within in-memory database

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
\ No newline at end of file
+## Introduction
+
+This example contains script *rdf4j-update.sms.ttl* that contain one pipeline. It creates RDF4J repository and performs update query. For more details about script construction and execution you can see [hello-world-example](https://github.com/kbss-cvut/s-pipes/blob/main/doc/examples/hello-world/hello-world.md).
```
```suggestion
This example explains script [rdf4j-update.sms.ttl](./rdf4j-update.sms.ttl]) that contain one pipeline. It creates RDF4J repository and performs update query. For more details about script construction and execution you can see [hello-world-example](../hello-world/hello-world.md).
```

blcham replied:
- provide links whenever possible
- use relative links instead so when we move repository/directory we do not need to refactor

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+        sm:next :update ;
+    .
+
+2) Perform an update on repository using `a kbss:rdf4j-update` module. Repository is defined in the same way as in `:create-repository`. Update query is set by string in `sp:text` section.
```
```suggestion
2) Perform an update on repository using `kbss:rdf4j-update` module. Repository is defined in the same way as in `:create-repository`. Update query is set by string in `sp:text` section.
```

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+            a sp:Update ;
+            sp:text """
+    DELETE {
+        <http://example.org/people/john>  <http://example.org/people/age> ?oldAge .
```
Please define a prefix in the script:
`ex-people: http://example.org/people/`

so this si more readable and we demonstrate and test that prefixes are inherited from script file.

rodionnv replied:
There is a problem with defining a prefix:
I have included it in the script like the other prefixes:
`@prefix ex-people: <http://example.org/people/> .`

but I get this error:
```
03-07-2023 18:04:34.413 [http-nio-8080-exec-10] ERROR c.c.s.m.Rdf4jUpdateModule - Update execution exception, query text:

DELETE {
	ex-people:john ex-people:age ?oldAge .
}
INSERT {
	ex-people:john ex-people:age ?newAge .
} WHERE {
   OPTIONAL {
      ex-people:john ex-people:age ?oldAge .
   }
   BIND(COALESCE(?oldAge+1, 1) as ?newAge)
}
      
org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName 'ex-people:john' uses an undefined prefix
```

blcham replied:
I suggest following: 
- add prefix inside of sp:text to fix the issue
- make new github issue to add this new feature
- add link of this new issue to the description of this issue (so the issues are interlinked)
- we discuss how to implement it on next meeting (put a note to the agenda please)

rodionnv replied:
Resolved

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+        ];
+        rdf4j:p-rdf4j-server-url "http://localhost:8080/rdf4j-server/" ;
+        rdf4j:p-rdf4j-repository-name "test-update" ;
+        rdfs:label "Make insert update" ;
```
```suggestion
```

blcham replied:
You do not have label in the previous module ... so I suggest omitting it here as well.

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+2) Perform an update on repository using `a kbss:rdf4j-update` module. Repository is defined in the same way as in `:create-repository`. Update query is set by string in `sp:text` section.
+
+
+    :update
```
```suggestion
    :update-repository
```
The previous module you called create-repository, so I would call this update-repository.

Note, that you must update it in the script as well.

blcham replied:
In case there is collision in names, use something like:
`:update-repository-default-context`

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+
+    [http-nio-8080-exec-7] INFO  c.c.s.e.ExecutionEngineImpl -  ##### create-repository
+    ...
+    INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jCreateRepositoryModule.executeSelf Server url: http://localhost:8080/rdf4j-server/ Repository name: test-update Ignore flag: true
```
```suggestion
    INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jCreateRepositoryModule.executeSelf Server url: http://localhost:8080/rdf4j-server/, Repository name: test-update, Ignore if repository exists: true
```

blcham replied:
Note it should be fixed in code as well.

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+    ...
+    [http-nio-8080-exec-7] INFO  c.c.s.e.ExecutionEngineImpl -  ##### Make insert update
+    ...
+    INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.loadConfiguration Connection created for repository test-update
```
```suggestion
    DEBUG [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.loadConfiguration Connected to repository test-update
```

blcham commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -2,4 +2,70 @@
+    [http-nio-8080-exec-7] INFO  c.c.s.e.ExecutionEngineImpl -  ##### Make insert update
+    ...
+    INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.loadConfiguration Connection created for repository test-update
+    INFO [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.makeUpdate Update executed
```
```suggestion
    DEBUG [http-nio-8080-exec-7] cz.cvut.spipes.modules.Rdf4jUpdateModule.makeUpdate Update successful
```

blcham reviewed (commented):
Great work done. See my suggestions (in case you agree with them, you can approve/reject it directly from github. Note that some of them need modification of code as well)

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jCreateRepositoryModule.java:
```diff
@@ -11,10 +11,11 @@
 
 public class Rdf4jCreateRepositoryModule extends AbstractModule {
-    private static final Logger logger = Logger.getLogger(Rdf4jUpdateModule.class.getName());
+    private static final Logger logger = LoggerFactory.getLogger(Rdf4jUpdateModule.class.getName());
```
Look how it is done elsewhere in the code ... not sure but i believe we call it only LOG, not logger ... we should generate similar code in all modules as much as possible

blcham replied:
i mean `private static final Logger LOG = ...`

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jCreateRepositoryModule.java:
```diff
@@ -63,6 +64,7 @@ ExecutionContext executeSelf() {
 
         RepositoryManager repositoryManager = new RemoteRepositoryManager(rdf4jServerURL);
         repositoryManager.init();
+        logger.info("Server url: "+rdf4jServerURL+", Repsitory name: "+rdf4jRepositoryName+ ", Ignore if repository exist: "+rdf4jIgnoreIfExists);
```
Use this format, is much more readable:

```       
        LOG.info("Server url:{}, Repsitory name:{}, Ignore if repository exist:{}.",
            rdf4jServerURL,
            rdf4jRepositoryName
            rdf4jIgnoreIfExists);
```

blcham replied:
this is much more readable, so you can check the sentence easily.

rodionnv replied:
This one is resolved

blcham commented on doc/examples/rdf4j-update/rdf4j-update.sms.ttl:
```diff
@@ -50,6 +50,8 @@ INSERT {
       """ ;
     ];
+  kbss:has-max-iteration-count 5 ;
+  kbss:only-if-triple-count-changes false;
```
please rename to: `stop-iteration-on-stable-triple-count`

blcham replied:
default should be false as you have now.

rodionnv replied:
Done

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jUpdateModule.java:
```diff
@@ -77,7 +87,12 @@ ExecutionContext executeSelf() {
             for (Resource updateQueryResource : updateQueries) {
                 String updateQuery = updateQueryResource.getProperty(SP.text).getLiteral().getString();
-                makeUpdate(updateQuery, updateConnection);
+                for(int i = 0;i < iterationCount;i++) {
```
No, you are doing it opposite way around, correct is:

```
for(int i = 0;i < iterationCount;i++) {
    for (Resource updateQueryResource : updateQueries) {
        // update
     }
     long newTriplesCount = updateConnection.size();
     if(onlyIfTripleCountChanges && (newTriplesCount == oldTriplesCount) )break;
}
```

rodionnv replied:
Done

blcham commented:
@rodionnv changes are required, see my comments.

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jUpdateModule.java:
```diff
@@ -138,7 +142,7 @@ public void loadConfiguration() {
                 ,iterationCount
                 ,onlyIfTripleCountChanges);
-        updateRepository = new SPARQLRepository(rdf4jServerURL + "repositories/" + rdf4jRepositoryName + "/statements");
+        if(updateRepository == null)setUpdateRepository(new SPARQLRepository(rdf4jServerURL + "repositories/" + rdf4jRepositoryName + "/statements"));
```
very strange ... do not understand why you are doing so.

blcham replied:
@blcham look at executeSelf method if it load configuration

rodionnv commented on doc/examples/rdf4j-update/rdf4j-update.md:
```diff
@@ -0,0 +1,70 @@
+            a sp:Update ;
+            sp:text """
+    DELETE {
+        <http://example.org/people/john>  <http://example.org/people/age> ?oldAge .
```
Done

rodionnv commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jCreateRepositoryModule.java:
```diff
@@ -0,0 +1,97 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Rdf4jCreateRepositoryModule extends AbstractModule {
```
Done

rodionnv commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jCreateRepositoryModule.java:
```diff
@@ -0,0 +1,97 @@
+    private String rdf4jRepositoryName;
+
+    /**
+     * Don't try to create new repository if it already exists
```
Done

blcham commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jUpdateModule.java:
```diff
@@ -0,0 +1,148 @@
+    private List<Resource> updateQueries;
+
+    private Repository updateRepository;
+    private Update prepareUpdate;
```
It does not make sense to have it here

blcham replied:
resolved

blcham commented:
@rodionnv I implemented following:
- simplified tests: `execute` --> `executeSelf`
- simplified configuration of modules, where you can assign queries in String` instead of `Resource`
- changed strange condition in loadConfiguration

blcham commented:
@rodionnv congrats !, merging pull request
